source/plugins/go/src/ingestion_token_utils.go

```diff
@@ -0,0 +1,430 @@
+	} `json:"configurations"`
+}
+
+type IngestionTokenResponse struct {
+	Configurationid    string `json:"configurationId"`
+	Ingestionauthtoken string `json:"ingestionAuthToken"`
+}
+
+func getAccessTokenFromIMDS() (string, int64, error) {
+	Log("Info getAccessTokenFromIMDS: start")	
+	useIMDSTOkenProxyEndPoint := os.Getenv("USE_IMDS_TOKEN_PROXY_END_POINT")
+	imdsAccessToken := ""
+	var responseBytes []byte
+	var err error
+	
+	if (useIMDSTOkenProxyEndPoint != "" && strings.Compare(strings.ToLower(useIMDSTOkenProxyEndPoint), "true") == 0) {
+		Log("Info Reading IMDS Access Token from IMDS Token proxy endpoint")
+		MCS_ENDPOINT := os.Getenv("MCS_ENDPOINT")
+		var msi_endpoint *url.URL
+		msi_endpoint, err := url.Parse("http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https://" + MCS_ENDPOINT + "/")
+		if err != nil {
+			Log("getAccessTokenFromIMDS: Error creating IMDS endpoint URL: %s", err.Error())
+			return imdsAccessToken, 0, err
+		}
+		req, err := http.NewRequest("GET", msi_endpoint.String(), nil)
+		if err != nil {
+			Log("getAccessTokenFromIMDS: Error creating HTTP request: %s", err.Error())
+			return imdsAccessToken, 0, err
+		}
+		req.Header.Add("Metadata", "true")
+
+		// Call managed services for Azure resources token endpoint
+		var resp *http.Response = nil
+		for i := 0; i < 3; i++ {		
+			resp, err = HTTPClient.Do(req)
+			if err != nil {
+				message := fmt.Sprintf("getAccessTokenFromIMDS: Error calling token endpoint: %s", err.Error())
+				Log(message)
+				SendException(message) // send the exception here because this error is not returned. The calling function will send any returned errors to telemetry.
+				continue
+			}
+
+			if resp != nil && resp.Body != nil {
+			  defer resp.Body.Close() 
+			}
+
+			Log("getAccessTokenFromIMDS: IMDS Response Status: %d", resp.StatusCode)
+			if resp.StatusCode != 200 {
+				message := fmt.Sprintf("getAccessTokenFromIMDS: IMDS Request failed with an error code : %d", resp.StatusCode)
+				Log(message)
+				SendException(message)
+				continue
+			}
+			break // call succeeded, don't retry any more
+		}
+		if resp == nil {
+			Log("getAccessTokenFromIMDS: IMDS Request ran out of retries")
+			return imdsAccessToken, 0, err
+		}
+
+		// Pull out response body
+		responseBytes, err = ioutil.ReadAll(resp.Body)
+		if err != nil {
+			Log("getAccessTokenFromIMDS: Error reading response body : %s", err.Error())
+			return imdsAccessToken, 0, err
+		}
+
+	} else {
+		Log("Info Reading IMDS Access Token from file : %s", IMDSTokenPathForWindows)
+		if _, err = os.Stat(IMDSTokenPathForWindows); os.IsNotExist(err) {
+			Log("getAccessTokenFromIMDS: IMDS token file doesnt exist: %s", err.Error())
+			return imdsAccessToken, 0, err
+		}
+		//ensure this works and no issues accessing the file
+		responseBytes, err = ioutil.ReadFile(IMDSTokenPathForWindows)
+		if err != nil {
+			Log("getAccessTokenFromIMDS: Could not read IMDS token from file: %s", err.Error())
```

Comment:
Should we have retries here? (in case there's an error reading the file, maybe kubernetes is updating it at the same time?)

Reply:
added retries. for public preview, we will be moving to IMDS endpoint.